### PR TITLE
[FIX] web: Get the correct list of actions on hr.leave form view

### PR DESCRIPTION
**Pull request by mamatch**
Steps to reproduce:
- Go to base Runbot and install Time Off.
- Go to Time Off > List View and select a list view item (checkbox).
- Go to Action and notice there is no Archive option.
- Go to the form view and go to Action > Archive.
- See error about not being able to archive leaves.

Issue:
We can see the action archive but we can not run it

Expected behaviour:
We shouldn't be able to see that action archive

Cause:
When computing the value of archiveEnabled we use the wrong variable "activeFields".

Solution:
Give the good one, "this.props.fields"

opw-3079475

### CODE REVIEW

**Review by nle-odoo — APPROVED** [review]
looks good to me

**Comment by mamatch** [comment]
I added the test :)

File: addons/web/static/tests/legacy/views/form_tests.js [hunk 1]
```diff
@@ -3034,6 +3034,26 @@ QUnit.module('LegacyViews', {
...
+            res_id: 1,
+            viewOptions: { hasActionMenus: true },
+            arch: '<form><field name="foo"/></form>',
+        });
+
+        assert.containsNone(form, '.o_cp_action_menus a:contains(Archive)');
+
+        form.destroy();
+    });
+
```
Comment (brboi) [inline addons/web/static/tests/legacy/views/form_tests.js:3056]: This test should be done in `tests/views/form_view_tests.js` instead.
Here you are testing the legacy form view.

cc @aab-odoo if you have any comment on this

**Review by brboi — CHANGES_REQUESTED** [review]

File: addons/web/static/src/views/form/form_controller.js [hunk 2]
```diff
@@ -153,9 +153,9 @@ export class FormController extends Component {
         // enable the archive feature in Actions menu only if the active field is in the view
         this.archiveEnabled =
             "active" in activeFields
-                ? !activeFields.active.readonly
+                ? !activeFields.active.modifiers.readonly
```
Comment (aab-odoo) [inline addons/web/static/src/views/form/form_controller.js:156]: In list, we read inside `fields` (`this.props.fields`) instead of `activeFields`. This is also what we did in legacy. So instead of reading inside `modifiers` here, I would replace `activeFields` by `this.props.fields` in this expression.

**Comment by mamatch** [comment]
Done :), Can you have a look please ?

File: addons/web/static/tests/views/form/form_view_tests.js [hunk 3]
```diff
@@ -12803,4 +12803,21 @@ QUnit.module("Views", (hooks) => {
...
+
+         await makeView({
+            type: "form",
+            resModel: "partner",
+            resId: 1,
+            data: this.data,
+            arch: `<form><field name="foo"/></form>`,
+        });
+
+        assert.containsNone(target, '.o_cp_action_menus a:contains(Archive)');
```
Comment (brboi) [inline addons/web/static/tests/views/form/form_view_tests.js:12820]: this assertion is false and does not have a safeguard if the action items become anything else than an `<a/>` tag.

You should instead:
- open the actions dropdown
- assert the list of action items

**Review by brboi — CHANGES_REQUESTED** [review]

File: addons/web/static/src/views/form/form_controller.js [hunk 4]
```diff
@@ -153,9 +153,9 @@ export class FormController extends Component {
         // enable the archive feature in Actions menu only if the active field is in the view
         this.archiveEnabled =
             "active" in activeFields
```
Comment (aab-odoo) [inline addons/web/static/src/views/form/form_controller.js:155]: here as well

File: addons/web/static/src/views/form/form_controller.js [hunk 5]
```diff
@@ -153,9 +153,9 @@ export class FormController extends Component {
...
-                ? !activeFields.active.readonly
+                ? !this.props.fields.active.readonly
                 : "x_active" in activeFields
```
Comment (aab-odoo) [inline addons/web/static/src/views/form/form_controller.js:157]: and here

File: addons/web/static/tests/views/form/form_view_tests.js [hunk 6]
```diff
@@ -12803,4 +12803,26 @@ QUnit.module("Views", (hooks) => {
...
+        await makeView({
+            type: "form",
+            resModel: "partner",
+            resId: 1,
+            serverData,
+            arch: `<form><field name="foo"/></form>`,
+            actionMenus: {},
+        });
+        await click(target.querySelector(".o_cp_action_menus"))
+        assert.containsNone(target, ".o_menu_item:contains(Archive)");
```
Comment (brboi) [inline addons/web/static/tests/views/form/form_view_tests.js:12825]: try something like this:
```suggestion
        await click(target, ".o_cp_action_menus");
        assert.deepEqual(
            [...target.querySelectorAll(".o_menu_item")].map((el) => el.textContent),
            ["action 1", "action 2"], // ...
            "should not contain an Archive action"
        );
```

**Review by brboi — CHANGES_REQUESTED** [review]

File: addons/web/static/tests/views/form/form_view_tests.js (hunk 6 above)
Comment (brboi) [reply addons/web/static/tests/views/form/form_view_tests.js:12825]: when you debug the test, inspect the dropdown and see how many menu items you should see.

Maybe "o_menu_item" is not the proper dom selector.

Comment (brboi) [reply addons/web/static/tests/views/form/form_view_tests.js:12825]: You should see this
image

Comment (mamatch) [reply addons/web/static/tests/views/form/form_view_tests.js:12825]: Exactly

Comment (brboi) [reply addons/web/static/tests/views/form/form_view_tests.js:12825]: replace selector
```
.o_cp_action_menus
``` 

with
```
.o_cp_action_menus .dropdown-toggle
```

**Comment by mamatch** [comment]
There is a test which fails and its not the one I wrote, do I modify it ? Can you have a look  please?

File: addons/web/static/src/views/form/form_controller.js [hunk 7]
```diff
@@ -152,10 +152,10 @@ export class FormController extends Component {
...
         // enable the archive feature in Actions menu only if the active field is in the view
         this.archiveEnabled =
-            "active" in activeFields
-                ? !activeFields.active.readonly
-                : "x_active" in activeFields
-                ? !activeFields.x_active.readonly
+            "active" in this.props.fields
+                ? !this.props.fields.active.readonly
+                : "x_active" in this.props.fields
+                ? !this.props.fields.x_active.modifiers.readonly
```
Comment (brboi) [inline addons/web/static/src/views/form/form_controller.js:158]: It looks like in a form view the archive action should never be there if the "active" field is not written in the arch.
@aab-odoo is it true ?

```suggestion
            "active" in activeFields
                ? !this.props.fields.active.readonly
                : "x_active" in activeFields
                ? !this.props.fields.x_active.readonly
```

File: addons/web/static/tests/views/form/form_view_tests.js [hunk 8]
```diff
@@ -12803,4 +12803,31 @@ QUnit.module("Views", (hooks) => {
...
+        });
+        await click(target, ".o_cp_action_menus .dropdown-toggle");
+        assert.deepEqual(
+            [...target.querySelectorAll(".o_menu_item")].map((el) => el.textContent),
+         ["Duplicate", "Delete"],
+            "Should not contain an Archive action",
+
+        );
+
+    });
```
Comment (brboi) [inline addons/web/static/tests/views/form/form_view_tests.js:12832]: In order to keep things a little tidy, move this new test right after the one named "archive action with active field not in view".

**Review by brboi — CHANGES_REQUESTED** [review]
Please make sure you apply the linting rules provided with the JS tooling: https://github.com/odoo/odoo/wiki/Javascript-coding-guidelines#tooling

File: addons/web/static/src/views/form/form_controller.js (hunk 7 above)
Comment (aab-odoo) [reply addons/web/static/src/views/form/form_controller.js:158]: Ho indeed. Otherwise we can't know whether we must display "Archive" or "Unarchive"

Comment (brboi) [reply addons/web/static/src/views/form/form_controller.js]: Just to make things clear, in list views we do not account for the `active` field being present in their archs. => expected behavior, isn't it ?

File: addons/web/static/src/views/form/form_controller.js [hunk 9]
```diff
@@ -153,9 +153,9 @@ export class FormController extends Component {
         // enable the archive feature in Actions menu only if the active field is in the view
         this.archiveEnabled =
             "active" in activeFields
-                ? !activeFields.active.readonly
+                ? !this.props.fields.active.readonly
                 : "x_active" in activeFields
-                ? !activeFields.x_active.readonly
+                ? !this.props.fields.x_active.modifiers.readonly
```
Comment (brboi) [inline addons/web/static/src/views/form/form_controller.js:158]: why look into the `modifiers`?

Comment (mamatch) [reply addons/web/static/src/views/form/form_controller.js:158]: Sorry just a missing of myself

File: addons/web/static/src/views/form/form_controller.js (hunk 7 above)
Comment (mamatch) [reply addons/web/static/src/views/form/form_controller.js]: Is it well done now ?

File: addons/web/static/tests/views/form/form_view_tests.js [hunk 10]
```diff
@@ -12803,4 +12830,5 @@ QUnit.module("Views", (hooks) => {
         await click(target.querySelector(".o_data_row .o_data_cell"));
         assert.containsOnce(target, ".o_form_view");
     });
+
```
Comment (brboi) [inline addons/web/static/tests/views/form/form_view_tests.js:12833]: unnecessary diff

File: addons/web/static/tests/views/form/form_view_tests.js [hunk 11]
```diff
@@ -4021,6 +4021,33 @@ QUnit.module("Views", (hooks) => {
...
+            resModel: "partner",
+            resId: 1,
+            serverData,
+            arch: `<form><field name="foo"/></form>`,
+            actionMenus: {},
+        });
+        await click(target, ".o_cp_action_menus .dropdown-toggle");
+        assert.deepEqual(
+            [...target.querySelectorAll(".o_menu_item")].map((el) => el.textContent),
+         ["Duplicate", "Delete"],
```
Comment (brboi) [inline addons/web/static/tests/views/form/form_view_tests.js:4044]: Please make sure you apply the linting rules provided with the JS tooling: https://github.com/odoo/odoo/wiki/Javascript-coding-guidelines#tooling

File: addons/web/static/tests/views/form/form_view_tests.js [hunk 12]
```diff
@@ -4021,6 +4021,33 @@ QUnit.module("Views", (hooks) => {
...
+            type: "char",
+            default: true,
+            readonly: true,
+        };
+        await makeView({
+            type: "form",
+            resModel: "partner",
+            resId: 1,
+            serverData,
+            arch: `<form><field name="foo"/></form>`,
```
Comment (brboi) [inline addons/web/static/tests/views/form/form_view_tests.js:4038]: add the field "active"

**Review by brboi — CHANGES_REQUESTED** [review]

File: addons/web/static/tests/views/form/form_view_tests.js (hunk 12 above)
Comment (mamatch) [reply addons/web/static/tests/views/form/form_view_tests.js:4038]: If I add the field "active" in the view, should I leave it in the partner model ?

File: addons/web/static/src/views/form/form_controller.js (hunk 7 above)
Comment (brboi) [reply addons/web/static/src/views/form/form_controller.js]: yes (y)

File: addons/web/static/tests/views/form/form_view_tests.js (hunk 12 above)
Comment (brboi) [reply addons/web/static/tests/views/form/form_view_tests.js:4038]: Why don't you try ?

Comment (mamatch) [reply addons/web/static/tests/views/form/form_view_tests.js]: I think it's ok now.

File: addons/web/static/tests/views/form/form_view_tests.js [hunk 13]
```diff
@@ -4021,6 +4021,30 @@ QUnit.module("Views", (hooks) => {
         assert.containsNone(target, ".o_cp_action_menus span:contains(Unarchive)");
     });
 
+    QUnit.debug("archive action not shown with readonly active field", async function (assert) {
```
Comment (brboi) [inline addons/web/static/tests/views/form/form_view_tests.js:4024]: Qunit.debug :-)

Comment (mamatch) [reply addons/web/static/tests/views/form/form_view_tests.js]: Sorry :/, done

**Review by brboi — APPROVED** [review]
Thanks for your work here ;-)

**Comment by mamatch** [comment]
Hi @brboi , Can I ahve a r+ here :)?

**Comment by brboi** [comment]
> Hi @brboi , Can I ahve a r+ here :)?

Hello @mamatch
The commit message does not match the changes, could you update it ?
Thanks

**Comment by mamatch** [comment]
Done :-)

**Review by brboi — APPROVED** [review]
robodoo r+
